src/main/java/hudson/plugins/s3/cloudfront/CloudFrontInvalidatePublisher.java

```diff
@@ -0,0 +1,188 @@
+import hudson.model.AbstractProject;
+import hudson.plugins.s3.S3BucketPublisher;
+import hudson.plugins.s3.S3Profile;
+import hudson.tasks.BuildStepDescriptor;
+import hudson.tasks.BuildStepMonitor;
+import hudson.tasks.Publisher;
+import hudson.tasks.Recorder;
+import hudson.util.CopyOnWriteList;
+import hudson.util.ListBoxModel;
+
+import java.io.IOException;
+import java.io.PrintStream;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+import jenkins.model.Jenkins;
+
+import org.apache.commons.lang.StringUtils;
+import org.kohsuke.stapler.DataBoundConstructor;
+
+public final class CloudFrontInvalidatePublisher extends Recorder implements Describable<Publisher> {
+	
+	@Extension
+	public static final DescriptorImpl DESCRIPTOR = new DescriptorImpl();
+
+	private String profileName;
+	private final List<InvalidationEntry> invalidationEntries;
+
+	public CloudFrontInvalidatePublisher(){
+		super();
+		this.invalidationEntries = Collections.emptyList();
+	}
+	
+	@DataBoundConstructor
+	public CloudFrontInvalidatePublisher(String profileName, List<InvalidationEntry> invalidationEntries) {
+		if (StringUtils.isBlank(profileName)) {
+			// defaults to the first one
+			S3Profile[] sites = DESCRIPTOR.getProfiles();
+			if (sites.length > 0)
+				profileName = sites[0].getName();
+		}
+
+		this.profileName = profileName;
+		this.invalidationEntries = invalidationEntries;
+
+	}
+
+	public String getProfileName() {
+		return profileName;
+	}
+
+	public List<InvalidationEntry> getInvalidationEntries() {
+		return invalidationEntries;
+	}
+
+	public S3Profile getProfile() {
+		return getProfile(profileName);
+	}
+
+	public static S3Profile getProfile(String profileName) {
+		S3Profile[] profiles = DESCRIPTOR.getProfiles();
+
+		if (profileName == null && profiles.length > 0)
+			// default
+			return profiles[0];
+
+		for (S3Profile profile : profiles) {
+			if (profile.getName().equals(profileName))
+				return profile;
+		}
+		return null;
+	}
+
+	protected void log(final PrintStream logger, final String message) {
+		logger.println(StringUtils.defaultString(getDescriptor().getDisplayName()) + " " + message);
+	}
+
+	@Override
+	public boolean perform(AbstractBuild<?, ?> build, Launcher launcher, BuildListener listener) throws InterruptedException, IOException {
+
+		final boolean buildFailed = build.getResult() == Result.FAILURE;
+
+		S3Profile profile = getProfile();
+		if (profile == null) {
+			log(listener.getLogger(), "No S3 profile is configured.");
+			build.setResult(Result.UNSTABLE);
+			return true;
+		}
+		log(listener.getLogger(), "Using S3 profile: " + profile.getName());
+		try {
+			Map<String, String> envVars = build.getEnvironment(listener);
+			for (InvalidationEntry entry : invalidationEntries) {
+
+				if (entry.noInvalidateOnFailure && buildFailed) {
+					// build failed. don't post
+					log(listener.getLogger(), "Skipping S3 key invalidation because build failed");
```

Comment:
That's true, i was looking at CloudFront in relation with S3. Actually, that's how we are using it. After we update some resources on S3, we want them to be updated within cdn as well, so we invoke invalidation on those keys. And because of the relation, we ask user to provide s3 bucket name and key prefix.
The callable, extracts all distributions related with the account and invalidate all distributions related to the bucket, so all of them have updated version of S3 file.

On the other hand, we can ask user to specify distribution alias instead of bucket and CloudFront object path prefix. @bsideup , do you think it would be more useful approach? What if user has multiple caches for same entries, would he/she has to create multiple invalidationEntries in the job config with same entry path?